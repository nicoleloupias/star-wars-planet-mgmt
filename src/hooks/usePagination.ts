import { useCallback, useMemo, useState } from "react";

export const usePagination = <T>(data?: T[], size: number = 10) => {
  const totalPages = useMemo(() => {
    if (!data) return 0;

    return Math.ceil(data.length / size);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [size, data?.length]);
  const [page, setPage] = useState(0);

  const onNextPage = useCallback(() => {
    setPage((prevState: number) => {
      if (prevState < totalPages) return prevState + 1;
      return prevState;
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [totalPages]);

  const onPrevPage = useCallback(() => {
    setPage((prevState: number) => {
      if (prevState > 0) return prevState - 1;
      return prevState;
    });
  }, []);

  const slicedData = useMemo(() => {
    if (!data) return data;
    return data.slice(page * size, page * size + size);
  }, [data, size, page]);
  const hasNextPage = useMemo(
    () => !(page === totalPages - 1),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [data, page]
  );
  const hasPrevPage = useMemo(
    () => !(page <= 0),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [data, page]
  );

  return {
    data: slicedData,
    page,
    totalPages,
    setPage,
    onPrevPage,
    onNextPage,
    hasNextPage,
    hasPrevPage
  };
};
