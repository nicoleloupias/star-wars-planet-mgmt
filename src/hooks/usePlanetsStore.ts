import { create } from "zustand";
import { Planet } from "../services/types/Planet";

interface PlanetStore {
  planets: Planet[];
  setPlanets: (planets: Planet[]) => void;
}

export const usePlanetsStore = create<PlanetStore>((set, get) => ({
  planets: [],
  setPlanets: (planets) => {
    return set((state) => {
      return { ...state, planets: [...state.planets, ...planets] };
    });
  }
}));
