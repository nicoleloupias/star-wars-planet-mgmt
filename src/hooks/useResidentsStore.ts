import { create } from "zustand";
import type { Planet } from "./usePlanetsStore";
import type { Resident } from "../services/types/Resident";

interface ResidentsStore {
  residents: Record<Planet["id"], Resident[]>;
  setResident: (planetId: Planet["id"], resident: Resident) => void;
}

export const useResidentsStore = create<ResidentsStore>((set) => ({
  residents: {},
  setResident: (planetId, resident) => {
    return set((state) => {
      return {
        ...state,
        residents: {
          ...state.residents,
          [planetId]: [...(state.residents?.[planetId] || []), resident]
        }
      };
    });
  }
}));
