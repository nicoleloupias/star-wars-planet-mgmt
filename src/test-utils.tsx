import { ChakraProvider } from "@chakra-ui/react";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import React from "react";
import { Router, RouterProvider } from "react-router-dom";
import { theme } from "./theme";
const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <ChakraProvider theme={theme}>
      <Router location={history.location} navigator={history}>
        {children}
      </Router>
    </ChakraProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {}
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => <AllProviders history={history}>{children}</AllProviders>,
    ...options
  });

export { customRender as render };
