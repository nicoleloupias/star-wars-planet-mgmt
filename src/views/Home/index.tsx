import { useEffect, useState } from "react";
import { PlanetService } from "../../services/PlanetService";
import { Box, Button, Center, Grid, GridItem, SimpleGrid } from "@chakra-ui/react";
import { PlanetCard } from "../../components/PlanetCard";
import { usePlanetsStore } from "../../hooks/usePlanetsStore";

export const Home = () => {
  const planets = usePlanetsStore((state) => state.planets);
  const setPlanets = usePlanetsStore((state) => state.setPlanets);
  const getAllPlanets = async () => {
    const planets = await PlanetService.getAll();

    setPlanets(planets);
  };

  useEffect(() => {
    if (planets.length === 0) {
      getAllPlanets();
    }
  }, []);

  return (
    <Center>
      <SimpleGrid mt={20} columns={{ base: 1, md: 2, lg: 3 }} gap={10}>
        {planets.length > 0 && planets?.map((planet) => <PlanetCard key={planet.url} {...planet} />)}
      </SimpleGrid>
    </Center>
  );
};
