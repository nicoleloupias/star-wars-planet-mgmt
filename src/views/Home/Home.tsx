import { Box, Center, Flex, SimpleGrid, Text } from "@chakra-ui/react";
import { PlanetCard } from "./components/PlanetCard";
import type { Planet } from "../../hooks/usePlanetsStore";
import { usePlanetsStore } from "../../hooks/usePlanetsStore";
import { Hero } from "./components/Hero";
import { useEffect, useState } from "react";
import { useForm } from "react-hook-form";
import { SortBy } from "./components/SortBy";
import { SearchBar } from "./components/SearchBar";

export interface FormFieldValues {
  search: string;
  sortBy: "name" | "diameter" | "climate" | "terrain" | "population";
}

export const Home = () => {
  const planets = usePlanetsStore((state) => state.planets);
  const [planetsToShow, setPlanetsToShow] = useState<Planet[] | undefined>();
  const form = useForm<FormFieldValues>();

  useEffect(() => {
    if (planets) {
      setPlanetsToShow([...planets]);
    }
  }, [planets]);

  return (
    <>
      <Hero />
      <Center flexDir="column">
        <Box as="section" minW={{ lg: "1040px" }} mt={10}>
          <Flex as="form" gap={2} justifyItems="flex-start" flexDir={{ base: "column", lg: "row" }}>
            <SearchBar setPlanetsToShow={setPlanetsToShow} form={form} />
            <SortBy setPlanetsToShow={setPlanetsToShow} planetsToShow={planetsToShow} form={form} />
          </Flex>

          <SimpleGrid my={20} columns={{ base: 1, md: 2, lg: 3 }} gap={10} justifyItems="center">
            {planetsToShow && planetsToShow?.map((planet) => <PlanetCard key={planet.id} {...planet} />)}
            {form.watch("search") && planetsToShow?.length === 0 && <Text>There are no results with this search</Text>}
          </SimpleGrid>
        </Box>
      </Center>
    </>
  );
};
