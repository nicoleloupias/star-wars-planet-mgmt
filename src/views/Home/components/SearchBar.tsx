import { Box, Flex, IconButton } from "@chakra-ui/react";
import { InputController } from "../../../components/InputController";
import { CloseIcon, SearchIcon } from "@chakra-ui/icons";
import type { Planet } from "../../../hooks/usePlanetsStore";
import { usePlanetsStore } from "../../../hooks/usePlanetsStore";
import type { UseFormReturn } from "react-hook-form";
import { FormFieldValues } from "../Home";

export interface SearchBarProps {
  setPlanetsToShow: React.Dispatch<React.SetStateAction<Planet[] | undefined>>;
  form: UseFormReturn<FormFieldValues>;
}

export const SearchBar = ({ form, setPlanetsToShow }: SearchBarProps) => {
  const { watch, control, resetField } = form;
  const planets = usePlanetsStore((state) => state.planets);
  const searchValue = watch("search");
  const handleSearchPlanet = (e: React.MouseEvent<HTMLElement>) => {
    e.preventDefault();
    const result = planets?.filter((planet) => {
      if (!searchValue) return planet;

      const search = searchValue.toLowerCase();

      return (
        planet.name.toLowerCase().includes(search) ||
        planet.climate.toLowerCase().includes(search) ||
        planet.terrain.toLowerCase().includes(search)
      );
    });

    setPlanetsToShow(result);
  };

  const handleResetSearch = () => {
    resetField("search");
    setPlanetsToShow([...(planets ?? [])]);
  };
  return (
    <Box maxW={{ lg: 60 }} w="full">
      <InputController
        id="search"
        showLabel={false}
        label="Search"
        control={control}
        iconRight={
          <Flex gap={2} mr={6}>
            {watch("search") && (
              <IconButton
                onClick={handleResetSearch}
                minW={4}
                size="xs"
                variant="transparent"
                aria-label="Remove search"
                icon={<CloseIcon />}
              />
            )}
            <IconButton
              onClick={handleSearchPlanet}
              minW={4}
              type="submit"
              size="xs"
              variant="transparent"
              aria-label="Search"
              icon={<SearchIcon />}
            />
          </Flex>
        }
      />
    </Box>
  );
};
