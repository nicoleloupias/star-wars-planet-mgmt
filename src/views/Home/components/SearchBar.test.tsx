import { beforeAll, describe, expect, test, vi } from "vitest";
import { render } from "../../../test-utils";
import { SearchBar } from "./SearchBar";
import { useForm } from "react-hook-form";
import { FormFieldValues } from "../Home";
import { usePlanetsStore } from "../../../hooks/usePlanetsStore";
import { fireEvent, waitFor } from "@testing-library/dom";
import { act } from "@testing-library/react";

describe("<SearchBar/>", () => {
  const setPlanetsToShow = vi.fn();
  const RenderWithForm = () => {
    const localForm = useForm<FormFieldValues>({
      mode: "onChange",
      criteriaMode: "all"
    });

    return <SearchBar form={localForm} setPlanetsToShow={setPlanetsToShow} />;
  };
  const factoryComponent = () => render(<RenderWithForm />);

  beforeAll(() => {
    usePlanetsStore.setState(
      {
        planets: [
          {
            id: "11",
            name: "Geonosis",
            rotation_period: "30",
            orbital_period: "256",
            diameter: "11370",
            climate: "temperate, arid",
            gravity: "0.9 standard",
            terrain: "rock, desert, mountain, barren",
            surface_water: "5",
            population: "100000000000",
            residents: ["https://swapi.dev/api/people/63/"],
            films: ["https://swapi.dev/api/films/5/"],
            created: "2014-12-10T12:47:22.350000Z",
            edited: "2014-12-20T20:58:18.437000Z",
            url: "https://swapi.dev/api/planets/11/"
          },
          {
            id: "12",
            name: "Utapau",
            rotation_period: "27",
            orbital_period: "351",
            diameter: "12900",
            climate: "temperate, arid, windy",
            gravity: "1 standard",
            terrain: "scrublands, savanna, canyons, sinkholes",
            surface_water: "0.9",
            population: "95000000",
            residents: ["https://swapi.dev/api/people/83/"],
            films: ["https://swapi.dev/api/films/6/"],
            created: "2014-12-10T12:49:01.491000Z",
            edited: "2014-12-20T20:58:18.439000Z",
            url: "https://swapi.dev/api/planets/12/"
          }
        ]
      },

      true
    );
  });

  test("should render succesfully", () => {
    const component = factoryComponent();

    expect(component).toMatchSnapshot();
    expect(component.getByTestId("search")).toBeInTheDocument();
  });

  test("should search succesfully", async () => {
    const component = factoryComponent();

    act(() => {
      fireEvent.change(component.getByTestId("search"), { target: { value: "Geonosis" } });
      fireEvent.click(component.getByRole("button", { name: "Search" }));
    });

    await waitFor(() => {
      expect(setPlanetsToShow).toHaveBeenCalledWith([
        {
          id: "11",
          name: "Geonosis",
          rotation_period: "30",
          orbital_period: "256",
          diameter: "11370",
          climate: "temperate, arid",
          gravity: "0.9 standard",
          terrain: "rock, desert, mountain, barren",
          surface_water: "5",
          population: "100000000000",
          residents: ["https://swapi.dev/api/people/63/"],
          films: ["https://swapi.dev/api/films/5/"],
          created: "2014-12-10T12:47:22.350000Z",
          edited: "2014-12-20T20:58:18.437000Z",
          url: "https://swapi.dev/api/planets/11/"
        }
      ]);
    });
  });
});
