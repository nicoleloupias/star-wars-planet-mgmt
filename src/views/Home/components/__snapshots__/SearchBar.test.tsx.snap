// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SearchBar/> > should render succesfully 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-dark"
  >
    <div>
      <div
        class="css-nu2kwa"
      >
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <div
            class="chakra-input__group css-1y0e7gb"
            data-group="true"
          >
            <input
              class="chakra-input css-1rrwde8"
              data-testid="search"
              id="search"
              name="search"
              placeholder="Search"
              value=""
            />
            <div
              class="chakra-input__right-element css-1lds0jh"
            >
              <div
                class="css-1rs9p06"
              >
                <button
                  aria-label="Search"
                  class="chakra-button css-1ni11ii"
                  type="submit"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-icon css-onkibi"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="css-nu2kwa"
    >
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <div
          class="chakra-input__group css-1y0e7gb"
          data-group="true"
        >
          <input
            class="chakra-input css-1rrwde8"
            data-testid="search"
            id="search"
            name="search"
            placeholder="Search"
            value=""
          />
          <div
            class="chakra-input__right-element css-1lds0jh"
          >
            <div
              class="css-1rs9p06"
            >
              <button
                aria-label="Search"
                class="chakra-button css-1ni11ii"
                type="submit"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-icon css-onkibi"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
